### PR TITLE
feat(shared-ini-file-loader): refactor loadSharedConfigFiles into modular components

### PR DESCRIPTION
**Pull request by trivikr**
### Issue
Fixes: https://github.com/aws/aws-sdk-js-v3/issues/3019
Refs: https://github.com/aws/aws-sdk-js-v3/issues/3279

### Description
* Refactors loadSharedConfigFiles to break it into individual components.
* Caches contents of readFile for a path, and ensures that it's called only once in parallel calls.

### Testing
* Unit testing
* `yarn test:integration`
* `yarn test:integration:legacy`

<details>
<summary>Single client</summary>

Refs: https://github.com/aws/aws-sdk-js-v3/issues/3019#issuecomment-1030258502
* Before: 4 calls to readFile
* After: 2 calls to readFile

</details>

<details>
<summary>Two clients</summary>

Refs: https://github.com/aws/aws-sdk-js-v3/issues/3019#issuecomment-1030265863
* Before: 8 calls to readFile
* After: 2 calls to readFile

</details>

<details>
<summary>Client with an operation</summary>

Refs: https://github.com/aws/aws-sdk-js-v3/issues/3019#issuecomment-1030283827
* Before: 16 calls to readFile
* After: 2 calls to readFile

</details>

### Additional Context
This is the merge of prior attempts in:
* https://github.com/aws/aws-sdk-js-v3/pull/3281
* https://github.com/aws/aws-sdk-js-v3/pull/3282

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/shared-ini-file-loader/src/getHomeDir.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ * Get the HOME directory for the current runtime.
+ *
+ * @internal
+ */
+export const getHomeDir = (): string => {
+  const { HOME, USERPROFILE, HOMEPATH, HOMEDRIVE = `C:${sep}` } = process.env;
+
+  if (HOME) return HOME;
+  if (USERPROFILE) return USERPROFILE;
+  if (HOMEPATH) return `${HOMEDRIVE}${HOMEPATH}`;
```
Comment (adrai) [inline packages/shared-ini-file-loader/src/getHomeDir.ts:14]: I hope this will work also for non-windows environments.

Comment (trivikr) [reply packages/shared-ini-file-loader/src/getHomeDir.ts:14]: This code is [existing code](https://github.com/aws/aws-sdk-js-v3/blob/63d7f8c539e0fd782fa5bf997bd6ac2730e4bfda/packages/shared-ini-file-loader/src/index.ts#L111-L124), and the PR just refactors it into a modular component for better code management.

Please create a [bug report](https://github.com/aws/aws-sdk-js-v3/issues/new?assignees=&labels=bug%2C+needs-triage&template=---bug-report.md&title=) if it's not working for specific non-windows environment and we'll triage it.

**Comment by trivikr** [comment]
PRs for test failures were posted and merged in:
* https://github.com/aws/aws-sdk-js-v3/pull/3287
* https://github.com/aws/aws-sdk-js-v3/pull/3289

**Review by AllanZhengYP — APPROVED** [review]
Looks good! Thank you for taking care of it!

File: packages/shared-ini-file-loader/src/slurpFile.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+const { readFile } = fsPromises;
+
+type callbacks = { resolve: Function; reject: Function };
+type FileStatus = {
+  contents: string;
+  isReading: boolean;
+  requestQueue: callbacks[];
+};
+
+const fileStatusHash: { [key: string]: FileStatus } = {};
```
Comment (AllanZhengYP) [inline packages/shared-ini-file-loader/src/slurpFile.ts:13]: I didn't realize we need to queue the promises by the file path. Looks good!

**Comment by trivikr** [comment]
npm pack size comparison:

<details>
<summary>Before: 7.4 kB</summary>

```console
$ yarn clean && rm -rf *.tsbuildinfo && yarn build

$ npm pack
...
npm notice package size:  7.4 kB                                    
npm notice unpacked size: 23.4 kB
...
```

</details>

<details>
<summary>After: 5.4 kB</summary>

```console
$ yarn clean && rm -rf *.tsbuildinfo && yarn build

$ npm pack
...
npm notice package size:  5.4 kB                                    
npm notice unpacked size: 15.4 kB
...
```

</details>

**Pull request merged**